pkg/gpu/nvidia/metrics/metrics.go

```diff
@@ -0,0 +1,155 @@
+
+	// MemoryTotal reports the total memory available on the GPU.
+	MemoryTotal = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "memory_total",
+			Help: "Total memory available on the GPU in bytes",
+		},
+		[]string{"namespace", "pod", "container", "make", "accelerator_id", "model"})
+
+	// MemoryUsed reports GPU memory allocated.
+	MemoryUsed = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "memory_used",
+			Help: "Allocated GPU memory in bytes",
+		},
+		[]string{"namespace", "pod", "container", "make", "accelerator_id", "model"})
+
+	// AcceleratorRequests reports the number of GPU devices requested by the container.
+	AcceleratorRequests = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "accelerator_requests",
+			Help: "Number of accelerator devices requested by the container",
+		},
+		[]string{"namespace", "pod", "container", "resource_name"})
+)
+
+// MetricServer exposes GPU metrics for all containers in prometheus format on the specified port.
+type MetricServer struct {
+	collectionInterval  int
+	port                int
+	metricsEndpointPath string
+}
+
+func NewMetricServer(collectionInterval, port int, metricsEndpointPath string) *MetricServer {
+	return &MetricServer{
+		collectionInterval:  collectionInterval,
+		port:                port,
+		metricsEndpointPath: metricsEndpointPath,
+	}
+}
+
+// Start performs necessary initializations and starts the metric server.
+func (m *MetricServer) Start() error {
+	glog.Infoln("Starting metrics server")
+
+	err := nvml.Init()
+	if err != nil {
+		return fmt.Errorf("failed to initialize nvml: %v", err)
+	}
+
+	driverVersion, err := nvml.GetDriverVersion()
+	if err != nil {
+		return fmt.Errorf("failed to query nvml: %v", err)
+	}
+	glog.Infof("nvml initialized successfully. Driver version: %s", driverVersion)
+
+	err = DiscoverGPUDevices()
+	if err != nil {
+		return fmt.Errorf("failed to discover GPU devices: %v", err)
+	}
+
+	go func() {
+		http.Handle(m.metricsEndpointPath, promhttp.Handler())
+		err := http.ListenAndServe(fmt.Sprintf(":%d", m.port), nil)
+		if err != nil {
+			glog.Infof("Failed to start metric server: %v", err)
+		}
+	}()
+
+	go m.collectMetrics()
+	return nil
+}
+
+func (m *MetricServer) collectMetrics() {
+	t := time.NewTicker(time.Millisecond * time.Duration(m.collectionInterval))
+	defer t.Stop()
+
+	for {
+		select {
+		case <-t.C:
+			devices, err := GetDevicesForAllContainers()
+			if err != nil {
+				glog.Errorf("Failed to get devices for containers: %v", err)
+				continue
+			}
+			m.updateMetrics(devices)
+		}
+	}
+}
+
+func (m *MetricServer) updateMetrics(containerDevices map[ContainerID][]string) {
+	for container, devices := range containerDevices {
+		AcceleratorRequests.WithLabelValues(container.namespace, container.pod, container.container, gpuResourceName).Set(float64(len(devices)))
+
+		for _, device := range devices {
+			d, err := DeviceFromName(device)
```

Comment:
why not call `DeviceUtilzation()` here?

Reply:
We anyway need to get the device from the device name so that we can reference other fields like UUID and Model. At that point, DeviceUtilzation() function doesn't add any value, since we can directly call d.Status() once we have the device pointer.  I think we can just remove the DeviceUtilzation() method, since it's not needed.